### DISCUSSION TITLE
Performing FEA simulations one after another in a single session

### DISCUSSION THREAD

**Discussion by abarun22** [idaholab/moose#16332] [Q&A General] [answered]
Dear all,
I would like to run a multiple numer of in-elastic simulations one after the another in a single run of MOOSE with an option to revise the input stress/plastic strain hardening parameters in between the simulations. Essentially we would want to start the simlation with a rough estimate of the hardening parameters and revise them based on the state of the internal stresses in each of the converged FE simulation. We would like to continue this iterative behaviour until the input parameters are all optimized and that the stresses agree with its experimental counterpart. Is that something we could try to make it happen in MOOSE? At present it does it perfectly for one FEA simulation. And advise on how to run continuous simulations in a single session of MOOSE would be greatly apprecietaed? 

Please see the attached input file i am using at present? Essentially i would be required to revise the contents of CSV file in the 'PiecewiseLinear' block  after each successful run of FEA simulation.
Kind regards,
Arun
[PieceWise_test.zip](https://github.com/idaholab/moose/files/5615765/PieceWise_test.zip)

**Comment by aeslaughter** [comment]
This is possible but will require you to write a little code using the stochastic tools module. You will need to create a custom Sampler object that will produce a new set of parameters to be modified at each timestep. @somu15 Could you add a comment about what you have done?

https://mooseframework.inl.gov/modules/stochastic_tools/index.html

**Comment by somu15** [comment, marked as answer]
Yes, we can run sequential samples using MOOSE STM such that the next sample is influenced by the current one. There are three things you need to do:

- There will be Main.i and sub.i files. Main.i controls your Monte Carlo simulations and sub.i is your model. In the Main.i file, set `num_rows = 1` in your sampler object. You should also have an executioner block with `num_steps` equal to the number of samples you want to generate.
- In the Main.i file again, you need to have the `Vectorpostprocessors` block with all the results your Sampler object needs to know to modify the next simulation.
- In your Sampler class (i.e., `NewSampler.C` you create), use the vectorpostporcessor interface to access the results from the previous run and to modify the input sample accordingly for the next run.

I hope the above will allow you get to started. If you are having trouble, I can put together some example input and .C files for your reference.

**Comment by abarun22** [comment]
Hi Som,
Sorry that it took a while to complete my testing activities before i could
take up this work. I will go through the stochastic modules to get a clear
understanding of the technicalities needed to implement this iterative
simulation. Would be glad to hear from you on the example files which would
help me to further my code developments.
Kind regards,
Arun

**Comment by somu15** [comment]
Hi Arun,

We have a pull request for a sequential sampler called Metropolis (a Markov Chain Monte Carlo sampler): https://github.com/idaholab/moose/pull/16398

This PR has a Metropolis class which does the sequential sampling and also has input file as part of the tests. Maybe you would like to take a look at this.

Let me know if you have questions.

Som

**Comment by ngrilli** [comment]
Dear @abarun22 

For your optimization problem, you could couple MOOSE with the DAKOTA optimization software from Sandia.
It is quite easy to use,
you will need first to write a postprocessing script that calculates a residual (difference) between your simulation data
and experimental data, then you prepare a DAKOTA input file with your favourite optimization procedure,
then you introduce the variable parameter in the MOOSE input file (adding the ${} symbol if I don't remember wrong),
which are the ones you want to optimize.
DAKOTA will launch for you the MOOSE simulations one after the other and will change the parameters in the MOOSE input file
at each iteration.

Best Regards,
Nicolò Grilli
National University of Singapore

**Comment by hugary1995** [comment]
@lynnmunday Do you want to comment here. I believe this is exactly what you did for your inverse problem moose app.

**Reply by lynnmunday** [reply]
The inverse optimization app I'm working on would do this but it is not merged and has no documentation.  I think we will have this cleaned up and merged into moose by the end of March.

**Comment by abarun22** [comment]
Dear Som/ Nicolò,
Thanks for your suggestions. I tried executing an example model with the
STM and it seems like the sampling is performed essentially at each time
step, which will not work in our case. Ideally we would require a sample to
be composed of all the time steps, then perform the revision of parameters
and then continue with the next sample (with all time steps) and then
terminate this chain at a certain convergence criteria dictated by the
internal stress state.  I tried to mimic this behaviour by setting the
identifier 'execute_on' to 'final' but without much success. Also this
method requires the specification of sample count (num_rows) prior to the
analysis which may not be possible, as it is determined at run-time based
on the chain convergence. I would welcome your suggestions as to whether
anything could be done in this regard to achieve the intended behaviour.
Attached here please find the input files and the execution log for this
computation.

As  Nicolò suggested, I did take a look at the DAKOTA optimization module
and from the outset it seems like a possible approach to tackle such kind
of optimization problems with ease. I will take a closer look at the
functionalities, try example problems and see if we could implement a
prototype that mimics the existing optimization approach handled through
ANSYS.
Kind regards,
Arun

**Reply by aeslaughter** [reply]
The scenario should be possible with STM. The main file would use StochasticFullSolveMultiapp that run complete simulations. If the main uses Transient executioner the Sampler you create would build a new set of samples at each "timestep" (iteration) then run the complete sub-application simulations again, with new samples.

**Reply by somu15** [reply]
@abarun22 There is a Subset Simulation class on my branch which does exactly what Andew mentioned above. Here is the link to the class: https://github.com/somu15/moose/blob/Subset/modules/stochastic_tools/src/samplers/SubTest.C

Here is a test input file: https://github.com/somu15/moose/blob/Subset/modules/stochastic_tools/test/tests/samplers/subtest/subsetsim.i

We are testing this class presently and we will submit a PR with an example soon.

**Reply by abarun22** [reply]
Dear Andrew/Somu/Nicolo,
Thanks for your suggestions. I further analysed whether STM approach can be tailored to run iterative type of FE simulations. The sampling approach that we use here is applied only in a pseudo sense, since we are not using the standard stochastic methods to perturb the parameters. Instead they are done through the results from the actual FEA and the experimental results. This mandates a two way communication between the main and sub apps, and thus the parameter 'num_rows' is always maintained one suggesting an execution of single sample, every time a parameter set is revised. I tried an example test case to establish connection between the main and sub app, but could not perform the other way. Unfortunately the two way coupled approach goes around an infinite loop without executing the results. 

Alteratively, i completely avoided the sampling method and introduced conventional multi-app approach with two similar models sitting on the main as well as sub apps domain. The models can be executed one after the other with the parameter set revised after the convergence checks. From my view, this seems to be the more neater way of running iterative simulations as against the STM method which seemed a bit tedious. I had a similar experience running this model as i failed to introduce a two way communication which is extremely vital in enabling a successful iterative simulation methodology. I have attached here the  input/output files used in these cases for your reference. The proposed methodology is explained in the attached slides. I would welcome your ideas/suggestions as whether these methods could be realised practically with minimal coding effort. Also please advise me on how a complete coupling (reverse connection) could be achieved for these methods. Please note that the parameter set is currently stored in a csv file and hence any further revision of this set can be re-written on the same file, thus saving efforts to store it on to special data structures. Also please note that the yield stress parameter ('yield_stress' found in STM/sub.i) needs to be modified after each revision of parameter set. I wonder whether a parameter substitution can be used here to enable a continuous usage of input file with right parameters.

I am also open to DAKOTA-MOOSE interface to implement scheme of things, if it offers a convenient way of handling this coupled approach. 

[20012021.zip](https://github.com/idaholab/moose/files/5895990/20012021.zip)

**Comment by aeslaughter** [comment]
We just added the ability to dynamically change the samples and the resulting sup-application runs. So, you could create a Sampler that is informed by the output of any number of MultiApps for creating the next sample. But, if you are happy with the MultiApp approach you mentioned, then I wouldn't worry about.

https://github.com/idaholab/moose/pull/16862
